src/ir/module-utils.h

```diff
@@ -402,7 +402,17 @@ inline void
 collectSignatures(Module& wasm,
                   std::vector<Signature>& signatures,
                   std::unordered_map<Signature, Index>& sigIndices) {
-  using Counts = std::unordered_map<Signature, size_t>;
+  struct Counts : public std::unordered_map<Signature, size_t> {
+    void note(Signature sig) { (*this)[sig]++; }
+    void maybeNote(Type type) {
+      if (type.isRef()) {
+        auto heapType = type.getHeapType();
+        if (heapType.isSignature()) {
+          note(heapType.getSignature());
```

Comment:
IIUC, this will miss some more interesting cases. For example, in this module:

```webassembly
(module
  (type $g (func))
  (type $f (func (param (ref $g))))
  (func $foo (param (ref $f) (ref $f)))
)
```

This code will properly note `$f` twice, but it will not note `$g`. I guess in general this will require a type walk. I'll throw up a PR adding a type walking function I have written locally that you can use.

Reply:
I think it will note `$g`? Doesn't the recursive traversing from line 470 do that?

Reply:
I think it works too. The roundtripped output contains this, which matches `$g`'s structure:
```wat
(module
 (type $none_=>_none (func))
```

Reply:
oh right, I forgot about that code 👍 
`